### PR TITLE
WIP: cli.py type hints

### PR DESCRIPTION
**Pull request by Ben-Reg**
Disclaimer: Tackling this was a bit more ambitious than I thought it would be.  I tried to take my time and do research, but I'm assuming there will be some work still needed before it can be merged.

I'm in no hurry if we need to let this sit until you have time to look at it.  And of course, if you take a look and decide it just needs too much help, feel free to reject and I'll look for a less involved file to practice on before coming back to this one.

Notes:

- The `forward_if_not_current()` function wrapper gave me some trouble, but I think it's done in the recommended way.  A second opinion would be great.
- Also had trouble knowing how to type FakeDict, FakeStream and FakeStdin in a way that wouldn't cause errors
- I tried to add a short explanation in a comment whenever I had to use `type:ignore`

Thanks.

### CODE REVIEW

**Comment by Ben-Reg** [comment]
Just looking at the errors that came up from the automated tests.

- The black error seems to be from a different file - I assume it's OK to ignore?
- I'll take a closer look at the mypy errors, but most of them seem to be coming from imports from other untyped files.  I ignored these when doing my own tests by using `--follow-imports=error`
- I'll follow up with the codespell error to see what that's about

**Comment by thomasballinger** [comment]
Glancing at this, good work! I can take a longer look by Saturday, feel free to ask any questions before then though. For black, even if it isn't introduced in your PR go ahead and run black on that file to fix. If you install the dev dependencies. For mypy, we do actually want to fix these, the easiest time to add make sure consumers of this code are typed is is when you change the types of this file (and you might actually change the way you type this file to make these match). I'd recommend installing mypy and running it locally with the same flags we do in the CI command you're looking at.

It looks like codespell is just a spellchecker that found a word it thinks is mispelled.

**Comment by Ben-Reg** [comment]
Ok, sounds good.  I've started working through the mypy import errors.  There's a lot from repl.py that will need to be typed, so it may take a little while to finish.

**Comment by Ben-Reg** [comment]
Ok, I've made some progress, but ran into a few questions.  Mypy is pointing out a number of missing attributes in repl.py.  From what I can tell, it looks like it is assumed that these attributes should be provided by the subclass (and they are provided in cli.py).

I wasn't sure of how you would want me to handle this - do I create empty class variables and methods in repl.py to satisfy mypy?

Looks like the tests are highlighting some other issues as well.  I'll look into those.

**Comment by Ben-Reg** [comment]
Ok, I've made some progress, but ran into a few questions.  Mypy is pointing out a number of missing attributes in repl.py.  From what I can tell, it looks like it is assumed that these attributes should be provided by the subclass (and they are provided in cli.py).

I wasn't sure of how you would want me to handle this - do I create empty class variables and methods in repl.py to satisfy mypy?

Looks like the tests are highlighting some other issues as well.  I'll look into those.

File: bpython/repl.py [hunk 1]
```diff
@@ -50,7 +61,8 @@
     have_pyperclip = False
 
 from . import autocomplete, inspection, simpleeval
-from .config import getpreferredencoding
+from .cli import Statusbar
```
Comment (sebastinas) [inline bpython/repl.py:64]: No, you cannot do that here. This leads to circular imports.

File: bpython/repl.py [hunk 2]
```diff
@@ -349,7 +365,7 @@ def clear(self) -> None:
 
 
 class Interaction:
-    def __init__(self, config, statusbar=None):
+    def __init__(self, config: Config, statusbar: Optional[Statusbar] = None):
```
Comment (sebastinas) [inline bpython/repl.py:368]: Note that the `urwid` frontends passes a different `Statusbar`. This needs be solved differently, e.g. by moving to the respective frontends.

**Review by sebastinas — CHANGES_REQUESTED** [review]

**Comment by thomasballinger** [comment]
@Ben-Reg I've not had time to look at this yet, will try to dig in this weekend. Or you're up to look together sometime send me an email at me@ballingt.com, we could screenshare and talk through some of the tricky bits.

**Comment by Ben-Reg** [comment]
That's a generous offer, @thomasballinger.  I'll send you an email.

File: bpython/repl.py (hunk 2 above)
Comment (thomasballinger) [reply bpython/repl.py:368]: @sebastinas what do you mean by moving to the respective frontends here?

**Comment by Ben-Reg** [comment]
Well, it looks like I may not get all the way there tonight.  I was able to find and fix the two pytest errors that were coming up before (with help from @thomasballinger - thanks!), but now I seem to be having problems with the python 3.7 tests (3.8 tests are running just fine on my machine).

Looks like using `curses.window` doesn't work with 3.7

File: bpython/curtsiesfrontend/repl.py [hunk 3]
```diff
@@ -357,7 +370,9 @@ def __init__(
         )
         self.edit_keys = edit_keys.mapping_with_config(config, key_dispatch)
         logger.debug("starting parent init")
-        super().__init__(interp, config)
+        # interp is a subclass of repl.Interpreter, so it definitely,
+        # implements the methods of Interpreter!
+        super().__init__(cast("Interpreter", interp), config)
```
Comment (sebastinas) [inline bpython/curtsiesfrontend/repl.py:376]: So why not mark `interp` to be of type `repl.Interpreter`?

File: bpython/repl.py (hunk 2 above)
Comment (sebastinas) [reply bpython/repl.py]: `Interaction.statusbar` is never used in `bpython/repl.py`. That's an implementation detail of the different repls that leaked into the base repl. If a repl needs that statusbar, they can add the statusbar to their `Interaction`.

File: bpython/repl.py (hunk 1 above)
Comment (Ben-Reg) [reply bpython/repl.py]: Thanks - this is fixed now.

File: bpython/repl.py (hunk 2 above)
Comment (Ben-Reg) [reply bpython/repl.py]: @sebastinas, I'm not positive this is the direction you wanted me to go, but I modified this to import Statusbar from both `cli` and `urwid` and then typed the parameter like this: `statusbar: Union["cliStatusbar", "urwidStatusbar", None] = None'

Let me know if I'm getting warmer or colder. :)

File: bpython/curtsiesfrontend/repl.py (hunk 3 above)
Comment (Ben-Reg) [reply bpython/curtsiesfrontend/repl.py:375]: @sebastinas, I've reworked this to use `repl.Interpreter` instead.

**Comment by sebastinas** [comment]
Thanks, merged.

**Pull request closed without merging**